### ISSUE TITLE
Parallel Execution, Appium hangs on first command (after driver init)

### ISSUE THREAD

**Issue by gamesguru** [appium/appium#9570]
## The problem
Working fine on one device, but hangs forever on the first command with two or more.
Example commands that hang for 60 seconds.. is this because of parallel execution?  How to work around
- getDeviceSize()
- element.Click()
- findElementsById()

## Environment

* Appium 1.6.5
* Windows/Linux

## Link to Appium logs

https://gist.github.com/gamesguru/7b740c93ea0683962006765699fe87bb

## Code To Reproduce Issue [ Good To Have ]
```

 private void Bw_DoWork(object sender, DoWorkEventArgs e)
        {
            int i = 0;
            foreach (string serial in nativeFuncs.initialSerials)
            {
                nativeFuncs.statusDict[serial] = "loading";
                nativeFuncs.appiumLogDict[serial] = "...starting appium...\n";
                nativeFuncs.portDict[serial] = 4723 + i;
                comboBox1.Items.Add(serial);
                i++;
            }
            //comboBox1.SelectedIndex = 0;
            nativeFuncs.setDevStats();
                     

            //runs the main tests
            Parallel.For(0, nativeFuncs.initialSize, new ParallelOptions { MaxDegreeOfParallelism = 4 }, z =>
            {
                string s = nativeFuncs.initialSerials[z];
                pReader.cmd($"/c appium -a 127.0.0.1 -p {nativeFuncs.portDict[s].ToString()}", s, true);

                DesiredCapabilities capabilities = new DesiredCapabilities();
                capabilities.SetCapability("udid", s);                                  //serial
                capabilities.SetCapability("deviceName", nativeFuncs.devNameDict[s]);   //SM-J327P (aka productModel)
                capabilities.SetCapability("appPackage", appPackage);
                capabilities.SetCapability("appActivity", appActivity);
                capabilities.SetCapability(MobileCapabilityType.NoReset, true);
                capabilities.SetCapability(MobileCapabilityType.FullReset, false);

                nativeFuncs.cmdWorkerDict[s].RunWorkerAsync();
                nativeFuncs.selfLogDict[s] += "starting appium log..\n";
                nativeFuncs.selfLogDict[s] += $"on port {nativeFuncs.portDict[s].ToString()}..\n";
                nativeFuncs.statusDict[s] = "Starting";
                try { nativeFuncs.driverDict[s] = new AndroidDriver<AndroidElement>(new Uri("http://127.0.0.1:" + nativeFuncs.portDict[s].ToString() + "/wd/hub"), capabilities); }
                catch (Exception ex) { nativeFuncs.selfLogDict[s] += printE(ex); }
                AndroidDriver<AndroidElement> driver = nativeFuncs.driverDict[s];
                if (driver == null)
                {
                    nativeFuncs.appiumLogDict[s] += "\n\nAppium failed to connect..";
                    //nativeFuncs.s
                    return;
                }

                driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(7);

                //grabs screen size
                //nativeFuncs.hDict[s] = driver.Manage().Window.Size.Height;
                //nativeFuncs.wDict[s] = driver.Manage().Window.Size.Width;
                //nativeFuncs.yDict[s] = nativeFuncs.hDict[s] / 2;
                //nativeFuncs.xDict[s] = nativeFuncs.wDict[s] / 2;

                //nativeFuncs.selfLogDict[s] += $"Height: {nativeFuncs.hDict[s]}, Width: {nativeFuncs.wDict[s]}";
                //Thread.Sleep(700);

                nativeFuncs.selfLogDict[s] += "\nAppium started, logging into OneControl...";
                nativeFuncs.statusDict[s] = "First run";
                nativeFuncs.setDevStats();

                if (!nativeFuncs.devNameDict[s].Contains("OCTP"))
                    testMethods.pinScreen(s);

                testMethods.TestLighting(s);
            });
```

**Comment by mykola-mokhnach** [comment]
Try Appium 1.7.2-beta

**Comment by gamesguru** [comment]
appreciate the suggestion but same problem sir on that version of appium.  one device works brilliantly, two fails miserably.  As you can see, uiAutomator never even really executes the commands.  It just waits forever on the first set.  

Possibly using FindElementByAndroidUIAutomator("UiSelector().text(\"Lighting\")") is the issue?  It doesn't seem to be a valid search method when dealing with multiple devices.  See the end of both logs.  Alternatives to that?

Or possibly just a problem with using a parallel loop and background workers in this way?  Needs cleaner solution?

logs for 1 good device:
https://gist.github.com/gamesguru/648c1312825178fcb026203b601f2bc8

logs for 2 bad devices:
https://gist.github.com/gamesguru/d7ee8f154dee0f9b3a8d54f2fe69ee02

**Comment by mykola-mokhnach** [comment]
If you say these are two different devices then why do you pass two similar udid values to both sessions capabilities?

**Comment by gamesguru** [comment]
i don't, they are stored in a dictionary and very different values.  We are testing proprietary android devices (not strictly phones).  check the logs you will see one is consistently passed one serial, another is consistently passed another... the same goes for port number.  still, some cross talk is causing issues behind the scene, probably with node, adb, or uiautomator

I did get further with uiautomator2, but it ended up throwing this error at the first unhandled command and failed a few handled commands before that: -		$exception	{"An unknown server-side error occurred while processing the command. Original error: Could not proxy command to remote server. Original error: Error: read ECONNRESET"}	System.InvalidOperationException

**Comment by mykola-mokhnach** [comment]
What I see in the logs is that commands from both sessions are sent to the same device and it causes the freeze.

**Comment by mykola-mokhnach** [comment]
> [Appium] Capabilities:
> [Appium]   udid: '03501af696b500000000'
> [Appium]   deviceName: 'OCTP_7'

and

> [Appium] Capabilities:
> [Appium]   udid: '03501af696b500000000'
> [Appium]   deviceName: 'OCTP_5'

**Comment by gamesguru** [comment]
you are comparing two different tests.  It is a bug in my app, if you check an item in the list of devices (on my form) it just loads the first device's serial and doesn't care which device you picked. 

However if you pick two (any two) it loads the first two serials (out of however many).  Therefore, i agree this is a bug in the way my dictionaries are stored and thank you for helping me to identify it, but this itself won't fix the appium concerns.

**Comment by saikrishna321** [comment]
@gamesguru make sure you provide different udid, different _systemPort_ to each thread for running parallel tests successfully. 

This FW will help you for implementation details

https://github.com/saikrishna321/AppiumTestDistribution

Closing as use-side issue

**Comment by gamesguru** [comment]
guys i can see where your confusion is stemming from, but i absolutely am using unique ports and serials.  the bug was not relevant to the test and after fixing it i am still having the same issues with appium that you can maybe help me solve?  have tried setting noReset back to true and searching for this /tmp/instrument_socks file to delete (haven't found it).  other suggestions welcome

ie. EADDRINUSE, ECONNREST, and most recently:
System.InvalidOperationException: 'An unknown server-side error occurred while processing the command. Original error: Could not proxy command to remote server. Original error: Error: connect ECONNREFUSED 127.0.0.1:8203'


you can clearly see the serials are being assigned properly to the ports
[Appium]   port: 4723
[Appium]   udid: '86501a68f79700000000'
[Appium]   deviceName: 'OCTP_7'
[Appium]   appPackage: 'com.lci1.one'

[Appium]   port: 4724
[Appium]   udid: '03501af696b500000000'
[Appium]   deviceName: 'OCTP_5'
[Appium]   appPackage: 'com.lci1.one'

**Comment by mykola-mokhnach** [comment]
@vmaxim Might this problem be somehow related to the fix you did recently for port forwarding?

**Comment by vmaxim** [comment]
@mykola-mokhnach Can you point me to the fix?

**Comment by mykola-mokhnach** [comment]
Sorry, this was another commit: https://github.com/appium/appium-chromedriver/commit/b55de252692ed2eb4691e07638969d73607bfb3a

@imurchie Might it be that when we cleaning ports then other parallel sessions get killed eventually?

**Comment by mykola-mokhnach** [comment]
@gamesguru Have you eventually read https://github.com/appium/appium/blob/55760b7b9e1c494c6a1f22d152f0bd1fe676dae7/docs/en/advanced-concepts/parallel-tests.md ?

To me it looks like both sessions still try to use the same port and the tutorial explicitly states

> systemPort If you are using appium-uiautomator2-driver, set a different system port for each Appium instanceset with systemPort capability since sometimes there can be a port conflict if different ports aren't used, such as in this issue.

**Comment by gamesguru** [comment]
i tried a bootstrap port earlier today (passed with the address and port to appium command line), that seemed to help a bit.  but still getting errors, see the gist.
https://gist.github.com/gamesguru/9a998a244ac85bcc14d27e20727f6f74

systemPort is not recognized in C# framework?  what am i doing wrong?

[Appium] Creating new AndroidUiautomator2Driver (v0.6.5) session
[Appium] Capabilities:
[Appium]   udid: '86501a68f79700000000'
[Appium]   deviceName: 'OCTP_7'
[Appium]   appPackage: 'com.lci1.one'
[Appium]   appActivity: 'md514e745197d12febda9505d9609e0ec28.SplashScreen'
[Appium]   automationName: 'uiautomator2'
[Appium]   skipUnlock: true
[Appium]   newCommandTimeout: 180
[Appium]   systemPort: 8702
[Appium]   noReset: true
[Appium]   fullReset: false
[Appium]   platformName: 'Android'
[BaseDriver] The following capabilities were provided, but are not recognized by appium: systemPort.
[BaseDriver] Session created with session id: 4c668ad3-9af5-456b-ab14-c2222d84b3e8
[debug] [AndroidDriver] Getting Java version
[AndroidDriver] Java version is: 1.8.0_144

**Comment by anandbhayre2007** [comment]
I am also facing same issue. Did anyone find the solution?

**Comment by saikrishna321** [comment]
@anandbhayre2007 Try this framework https://github.com/saikrishna321/AppiumTestDistribution/

**Comment by gamesguru** [comment]
sadly i am growing more convinced by the day that appium does not perform well at all in long-term tests, or ones with more than one device.  The problem: if uiautomator or adb go down for just a couple seconds (which they do several times a day under normal testing), the appium session is forcibly terminated.

it really shines in those basic unit tests to see if anything is horribly broken, but even then you have to hand swap out devices one at a time or run them in serial not parallel.

there are more reliable ways to do this, using adb shell uiautomator dump, managed libraries and xml parsers.  uiautomator dump may take 3-4 seconds compared to appium's instant response, but there is a price to be paid for not having your tests crash 3 hours after you leave the office.  team appium has some problems to solve

**Comment by anandbhayre2007** [comment]
@saikrishna321: Thank you for providing the solution, but we have got our own framework which is working fine at Offshore in India, but somehow same scripts are not working in clients environment. First 4 -5  steps work perfectly fine in all the devices, but after 4-5 steps it works only on one device and on rest of the devices it got hanged. The step on which it is hanging is related to tapping, so we are changing the context from web 
[appium-server-logs_Failed.txt](https://github.com/appium/appium/files/1690096/appium-server-logs_Failed.txt)
[appium-server-logs_Passed.txt](https://github.com/appium/appium/files/1690097/appium-server-logs_Passed.txt)

to native and native to web. Out of 10 runs, once or twice it works perfectly fine. I checked it with appium version 1.6.5 - 1.7.2.
